test/test_topology.ml

```diff
@@ -0,0 +1,204 @@
+
+let balancing nodes ~vms =
+  (* We expect to use many NUMA nodes when we have more VMs,
+   * more elaborate checks could be done on how well balanced the VMs are across NUMA nodes.
+   * Though in this case the default would always be best since it uses all nodes.
+   * *)
+  let nodes_used = List.sort_uniq compare nodes |> List.length in
+  min vms nodes_used |> float
+
+let check_aggregate_costs_not_worse (default, next, _) ~vms =
+  let default = sum_costs default in
+  let next = sum_costs next in
+  cost_not_worse ~default next ;
+  let balancing_default = balancing default.nodes ~vms in
+  let balancing_next = balancing next.nodes ~vms in
+  let balancing_best = max balancing_next balancing_default in
+  Alcotest.(
+    check (float 1e-3) "Balancing could improve" balancing_best balancing_next) ;
+  if balancing_next > balancing_default then D.debug "Balancing has improved!"
+
+let test_allocate ?(mem = Int64.shift_left 1L 30) h ~vms () =
+  let memsize = Int64.shift_left 1L 34 in
+  let nodea = NUMA.nodes h |> Seq.map (fun _ -> memsize) |> Array.of_seq in
+  let nodes =
+    NUMA.nodes h |> List.of_seq
+    |> List.mapi (fun i n -> NUMA.resource h n ~memory:nodea.(i))
+    |> Array.of_list
+  in
+  D.debug "NUMA: %s" (Fmt.to_to_string NUMA.pp_dump h) ;
+  let cores = NUMA.all_cpus h |> CPUSet.cardinal in
+  let vm_cores = max 2 (cores / vms) in
+  List.init vms (fun i -> i + 1)
+  |> List.fold_left
+       (fun (costs_old, costs_new, plans) i ->
+         D.debug "Planning VM %d" i ;
+         let vm = NUMARequest.v ~memory:mem ~vcpus:vm_cores in
+         match Softaffinity.plan h nodes ~vm with
+         | None ->
+             Alcotest.fail "No NUMA plan"
+         | Some plan ->
+             D.debug "NUMA allocation succeeded for VM %d: %s" i
+               (Fmt.to_to_string CPUSet.pp_dump plan) ;
+             let usednodes =
+               plan |> CPUSet.elements
+               |> List.map (NUMA.node_of_cpu h)
+               |> List.sort_uniq compare |> List.to_seq
+             in
+             let available_mem =
+               usednodes
+               |> Seq.map (fun (NUMA.Node i) -> nodea.(i))
+               |> Seq.fold_left Int64.add 0L
+             in
+             Alcotest.(
+               check int64 "Enough memory available on selected nodes" mem
+                 (min available_mem mem)) ;
+             let costs_numa_aware =
+               vm_access_costs h plans (vm_cores, usednodes, plan)
+             in
+             let costs_default =
+               vm_access_costs h plans (vm_cores, NUMA.nodes h, NUMA.all_cpus h)
+             in
+             cost_not_worse ~default:costs_default costs_numa_aware ;
+             ( costs_default :: costs_old
+             , costs_numa_aware :: costs_new
+             , ((vm_cores, List.of_seq usednodes), plan) :: plans ))
+       ([], [], [])
+  |> check_aggregate_costs_not_worse ~vms
+
+let mem3 = Int64.div (Int64.mul 4L (Int64.shift_left 1L 34)) 3L
```

Comment:
would be nice to have a comment about how this relates to the default `mem` value - `Int64.shift_left 1L 30`